src/txn/txn_ckpt.c

```diff
@@ -988,6 +988,16 @@ __txn_checkpoint(WT_SESSION_IMPL *session, const char *cfg[])
     }
 
 err:
+
+    /*
+     * Save clock value marking end of checkpoint processing. Backup uses this to determine which
+     * checkpoints it must keep.
+     */
+    __wt_seconds(session, &time_finish);
+    /* Be defensive: time is only monotonic per session */
+    if (time_finish > conn->ckpt_finish_time)
+        conn->ckpt_finish_time = time_finish;
+
```

Comment:
Responding to your first bullet: If I understand correctly, this means that if we wind up with checkpoints on some but not all tables, then it's OK to remove those checkpoints during a backup.  This could happen if the call to `checkpoint_apply_to_dhandles(... wt_checkpoint_sync())` on line 991 fails part way through the set of handles.  Seems like a reasonable course of action, and I'll move this code as you suggest.

Reply:
What happens if checkpoint fails partway through is that the metadata is not updated. Therefore, those completed checkpoints are not able to be "found" from the metadata address. The data exists in the files, yes. That is basically the same thing as eviction, which is writing to files before checkpoints but what is written cannot be found until the checkpoint completes and the metadata is updated.